### ISSUE TITLE
Web hook may use wrong store for multi-store with non-unique subscriber email addresses

### ISSUE THREAD

**Issue by TroyPatteson** [mailchimp/mc-magento#230]
Apologies in advance for this long post!

Multi-store Magento can be configured to either share customer accounts between all stores via **System -> Configuration -> Customers -> Account Sharing Options** or keep the customer accounts separate for each website.

When accounts are not shared between stores and a customer subscribes to multiple stores using the same email address there will be multiple customer accounts for that address. While the core Magento newsletter subscriber table stores both subscriber email address and store ID, Magento does not actually properly support multi-store newsletter because the two methods to retrieve a Newsletter Subscriber model only search based on subscriber email:

```php
    public function loadByEmail($subscriberEmail)
    {
        $select = $this->_read->select()
            ->from($this->getMainTable())
            ->where('subscriber_email=:subscriber_email');

        $result = $this->_read->fetchRow($select, array('subscriber_email'=>$subscriberEmail));

        if (!$result) {
            return array();
        }

        return $result;
    }

    /**
     * Load subscriber by customer
     *
     * @param Mage_Customer_Model_Customer $customer
     * @return array
     */
    public function loadByCustomer(Mage_Customer_Model_Customer $customer)
    {
        $select = $this->_read->select()
            ->from($this->getMainTable())
            ->where('customer_id=:customer_id');

        $result = $this->_read->fetchRow($select, array('customer_id'=>$customer->getId()));

        if ($result) {
            return $result;
        }

        $select = $this->_read->select()
            ->from($this->getMainTable())
            ->where('subscriber_email=:subscriber_email');

        $result = $this->_read->fetchRow($select, array('subscriber_email'=>$customer->getEmail()));

        if ($result) {
            return $result;
        }

        return array();
    }
```

Those functions retrieve the record for whichever store the customer first subscribed to, not necessarily the subscriber record according to whichever store the customer is currently accessing.

Contrast this with say the Customer resource model's equivalent loadByEmail():

```php
    public function loadByEmail(Mage_Customer_Model_Customer $customer, $email, $testOnly = false)
    {
        $adapter = $this->_getReadAdapter();
        $bind    = array('customer_email' => $email);
        $select  = $adapter->select()
            ->from($this->getEntityTable(), array($this->getEntityIdField()))
            ->where('email = :customer_email');

        if ($customer->getSharingConfig()->isWebsiteScope()) {
            if (!$customer->hasData('website_id')) {
                Mage::throwException(
                    Mage::helper('customer')->__('Customer website ID must be specified when using the website scope')
                );
            }
            $bind['website_id'] = (int)$customer->getWebsiteId();
            $select->where('website_id = :website_id');
        }

        $customerId = $adapter->fetchOne($select, $bind);
        if ($customerId) {
            $this->load($customer, $customerId);
        } else {
            $customer->setData(array());
        }

        return $this;
    }
```

If the customer sharing configuration is _by website_ the search for customer based on email address includes the website ID. Why the Magento development team did not do the same for Newsletter Subscriber is a mystery.

The way to solve this is to install the free _Netgo_MultistoreNewsletter_ extension which is nothing more than a class rewrite of the _Mage_Newsletter_Model_Subscriber_ class to search using email address and current store ID when calling **loadByEmail()** or customer's email and store ID when calling **loadByCustomer()**.

The problem is that the Mailchimp web hook is based on the default store view URL which means that the current store will be the default store regardless of which list the subscriber belongs to.

This can be tested by creating an account for the default store and then a second one for another store with the same email address where each store syncs to a different Mailchimp list.

The web hook has Mailchimp list ID and subscriber email address available in the post data for all web hook requests. I believe the correct solution is to find the list of Magento store IDs that currently have Mailchimp enabled and are configured to sync to the Mailchimp list ID in the web hook data and then search for Newsletter Subscriber or Customer models using email address and those store IDs.

To do that I have patched **Helper/Data.php** to append the following three functions:

```php
    /**
     * Return the list of magento store IDs configured to synchronise to
     * the given mailchimp list ID.
     * @param string $listId Mailchimp List ID
     * @returns int[] $magentoStoreIds Magento store IDs that sync to this list.
     */
    public function getMagentoStoreIdsByListId($listId)
    {
        $storeIds = Mage::registry('mailchimp_store_ids_for_list_'.$listId);
        if ($storeIds === null) {
            $stores = Mage::app()->getStores();
            $storeIds = array();
            foreach ($stores as $storeId => $store) {
                if ($this->isMailChimpEnabled($storeId)) {
                    $storeListId = $this->getConfigValueForScope(Ebizmarts_MailChimp_Model_Config::GENERAL_LIST, $storeId);
                    if ($storeListId == $listId)
                        $storeIds[] = $storeId;
                }
            }
            Mage::register('mailchimp_store_ids_for_list_'.$listId, $storeIds);
        }

        return $storeIds;
    }

    /**
     * Return the newsletter subscriber model for the given subscriber email
     * address for magento stores subscribed to the given Mailchimp List ID.
     * @param string $listId Mailchimp list ID
     * @param string $email Subscriber email address.
     * @returns Mage_Newsletter_Model_Subscriber $subscriber
     */
    public function loadListSubscriber($listId, $email)
    {
        $subscriber = null;

        $storeIds = $this->getMagentoStoreIdsByListId($listId);
        if (count($storeIds) > 0)
            $subscriber = Mage::getModel('newsletter/subscriber')->getCollection()
                ->addFieldToFilter('store_id', array('in' => $storeIds))
                ->addFieldToFilter('subscriber_email', $email)
                ->getFirstItem();

        if (!$subscriber)
            $subscriber = Mage::getModel('newsletter/subscriber')
                ->loadByEmail($email);

        return $subscriber;
    }

    /**
     * Return the customer model for the given subscriber email
     * address for magento stores subscribed to the given Mailchimp List ID.
     * @param string $listId Mailchimp list ID
     * @param string $email Subscriber email address.
     * @returns Mage_Customer_Model_Customer|null $customer
     */
    public function loadListCustomer($listId, $email)
    {
        $customer = null;

        $storeIds = $this->getMagentoStoreIdsByListId($listId);
        if (count($storeIds) > 0) {
            $customer = Mage::getResourceModel('customer/customer_collection')
                ->addFieldToFilter('store_id', array('in' => $storeIds))
                ->addFieldToFilter('email', array('eq' => $email))
                ->getFirstItem();
            if ($customer !== null) {
                $customer = Mage::getModel('customer/customer')->load($customer->getId());;
            }
        }

        return $customer;
    }
```

And then patched **Model/ProcessWebhook.php** as follows:

In **_updateEmail()** replace:

```php
    protected function _updateEmail(array $data)
    {
        $old = $data['data']['old_email'];
        $new = $data['data']['new_email'];

        $oldSubscriber = $this->loadByEmail($old);
        $newSubscriber = $this->loadByEmail($new);
```

with

```php
    protected function _updateEmail(array $data)
    {
        $listId = $data['data']['list_id'];
        $old = $data['data']['old_email'];
        $new = $data['data']['new_email'];

        $oldSubscriber = Mage::helper('mailchimp')->loadListSubscriber($listId, $old);
        $newSubscriber = Mage::helper('mailchimp')->loadListSubscriber($listId, $new);
```

In **_clean()** replace:

```php
    protected function _clean(array $data)
    {
        //Delete subscriber from Magento
        $s = $this->loadByEmail($data['data']['email']);
```

with

```php
    protected function _clean(array $data)
    {
        //Delete subscriber from Magento
        $s = Mage::helper('mailchimp')->loadListSubscriber($data['data']['list_id'], $data['data']['email']);
```

In **_subscribe()** replace:

```php
    protected function _subscribe(array $data)
    {
        try {
            $subscriber = Mage::getModel('newsletter/subscriber')
                ->loadByEmail($data['data']['email']);
```

with

```php
    protected function _subscribe(array $data)
    {
        try {
            $subscriber = Mage::helper('mailchimp')->loadListSubscriber($data['data']['list_id'], $data['data']['email']);

```

In **_unsubscribe()** replace:

```php
    protected function _unsubscribe(array $data)
    {
        $subscriber = $this->loadByEmail($data['data']['email']);
        if (!$subscriber->getId()) {
            $subscriber = Mage::getModel('newsletter/subscriber')
                ->loadByEmail($data['data']['email']);
        }
```

with

```php
    protected function _unsubscribe(array $data)
    {
        $subscriber = Mage::helper('mailchimp')->loadListSubscriber($data['data']['list_id'], $data['data']['email']);
```

In **_profile()** replace:

```php
    protected function _profile(array $data)
    {
        $email = $data['data']['email'];
        $subscriber = $this->loadByEmail($email);
        $customerId = null;

        $customerCollection = Mage::getResourceModel('customer/customer_collection')
            ->addFieldToFilter('email', array('eq' => $email));
        if ($customerCollection->getSize() > 0) {
            $customerId = $customerCollection->getFirstItem()->getEntityId();
        }

        if ($customerId) {
            $toUpdate = Mage::getModel('customer/customer')->load($customerId);
            $toUpdate->setFirstname($data['data']['merges']['FNAME']);
            $toUpdate->setLastname($data['data']['merges']['LNAME']);
        } else {
            $toUpdate = $subscriber;
            $toUpdate->setSubscriberFirstname($data['data']['merges']['FNAME']);
            $toUpdate->setSubscriberLastname($data['data']['merges']['LNAME']);
        }

        $toUpdate->save();


    }
```

with

```php
    protected function _profile(array $data)
    {
        $listId = $data['data']['list_id'];
        $email = $data['data']['email'];
        $customer = Mage::helper('mailchimp')->loadListCustomer($listId, $email);
        if ($customer) {
            $toUpdate = $customer;
            $toUpdate->setFirstname($data['data']['merges']['FNAME']);
            $toUpdate->setLastname($data['data']['merges']['LNAME']);
        } else {
            $subscriber = Mage::helper('mailchimp')->loadListSubscriber($listId, $email);
            $toUpdate = $subscriber;
            $toUpdate->setSubscriberFirstname($data['data']['merges']['FNAME']);
            $toUpdate->setSubscriberLastname($data['data']['merges']['LNAME']);
        }

        $toUpdate->save();
    }
```

And finally remove function **loadByEmail()** which is no longer used.

By the way, **_profile()** updates the first and last name in the customer record but doesn't modify the corresponding subscriber record. I wonder if it should.

Note that I put the **loadListSubscriber()** and **loadListCustomer()** in the helper as they may be useful elsewhere such as writing scripts.

**Comment by centerax** [comment]
@TroyPatteson Thanks for the detailed explanation. Can you create a pull request so we can review this properly?

**Comment by centerax** [comment]
thanks @TroyPatteson I think we are missing the final step. Can you create the PR?

**Comment by centerax** [comment]
can we close this one @TroyPatteson ? It seems its merged on pr #246

**Comment by TroyPatteson** [comment]
Yes please.
